optimum/onnx/graph_transformations.py

```diff
@@ -79,3 +96,219 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
+
+    deduplicated_initializers = list()
+    deduplicated_name = list()
+
+    for initializer in all_initializers:
+        if name_sharing_dict[initializer.name] not in deduplicated_name:
+            deduplicated_name.append(name_sharing_dict[initializer.name])
+            initializer.name = name_sharing_dict[initializer.name]
+            deduplicated_initializers.append(initializer)
+
+    return deduplicated_initializers
+
+
+def merge_decoders(
+    decoder: ModelProto,
+    decoder_with_past: ModelProto,
+    graph_name: str = "merged",
+    producer_name: str = "optimum-onnx",
+    simplify_graph: bool = False,
+) -> ModelProto:
+    """
+    Fuses decoder ONNX model and decoder with past ONNX model into one ONNX model with if logic.
+
+    Args:
+        decoder (`~onnx.ModelProto`): Decoder ONNX model.
+        decoder_with_past (`~onnx.ModelProto`): Decoder with past ONNX model.
+        graph_name (`~str`): Name of the parent graph(graph of the control flow node).
+        producer_name (`~str`): Graph producer name.
+        simplify_graph (`~bool`): Whether to use onnx-simplifier to simplify subgraphs.
+
+    Returns:
+        `~onnx.ModelProto`: The fused decoder ONNX model.
+    """
+    if simplify_graph:
+        decoder = simplify_onnx_graph(decoder)
+        decoder_with_past = simplify_onnx_graph(decoder_with_past)
+
+    _unify_onnx_outputs(decoder, decoder_with_past)
+    all_inputs = _get_all_inputs([decoder, decoder_with_past])
+    deduplicated_initializers = _deduplicated_cross_model_initializers([decoder, decoder_with_past], suffix=graph_name)
+
+    # Make subgraphs
+    no_past_branch = onnx.helper.make_graph(
+        nodes=decoder.graph.node,
+        name="no_past",
+        inputs=[],
+        outputs=decoder.graph.output,
+        initializer=[],
+    )
+    with_past_branch = onnx.helper.make_graph(
+        nodes=decoder_with_past.graph.node,
+        name="with_past",
+        inputs=[],
+        outputs=decoder_with_past.graph.output,
+        initializer=[],
+    )
+
+    # Merge subgraphs with a `If` node
+    use_cache = onnx.helper.make_tensor_value_info(
+        name="use_cache",
+        elem_type=onnx.TensorProto.BOOL,
+        shape=[1],
+    )
+    if_node = onnx.helper.make_node(
+        "If",
+        inputs=["use_cache"],
+        outputs=[output.name for output in no_past_branch.output],
+        name="optimum::if",
+        then_branch=with_past_branch,
+        else_branch=no_past_branch,
+    )
+    merged_graph = onnx.helper.make_graph(
+        nodes=[if_node],
+        name=graph_name,
+        inputs=all_inputs + [use_cache],
+        outputs=no_past_branch.output,
+        initializer=deduplicated_initializers,
+    )
+    merged_model = onnx.helper.make_model(
+        merged_graph,
+        producer_name=producer_name,
+        opset_imports=[
+            onnx.helper.make_opsetid(
+                domain=onnx.defs.ONNX_DOMAIN,
+                version=_get_onnx_opset(decoder),
```

Comment:
what happens if decoder and decoder_with_past use a different opset? Is there a check for it?

Reply:
It will fail during the checker or during inference. Good point, I will add a check here.

Reply:
But not sure if we want to introduce another tool that overlaps a bit with what we have.

Reply:
Added!